### PR TITLE
Serialize minidumps in a streaming fashion.

### PR DESCRIPTION
**Pull request by hlinnaka**
Serialize minidumps in a streaming fashion.
    
Instead of having a fixed-size buffer to serialize minidumps to, refactor
the serialization functions to write to a stream. That simplifies the
serialization functions, as they no longer need to reserve space in the
buffer ahead of time (i.e. the UlpRequiredSpace() functions are gone),
reduces memory usage when dumping small queries, and makes it possible to
minidump large queries without running out of memory.
    
This removes the arbitrary 16 MB limit on minidump size. I'm not sure it's
a good idea to create multi-gigabyte minidumps in practice, but that's a
case of "if it hurts, don't do it". At least it's now possible, if you need
to.

### CODE REVIEW

File: libgpopt/src/optimizer/COptimizer.cpp [hunk 1]
```diff
@@ -199,28 +201,34 @@ COptimizer::PdxlnOptimize
...
+
+		CMinidumperUtils::GenerateMinidumpFileName(szFileName, GPOS_FILE_NAME_BUF_SIZE, ulSessionId, ulCmdId, szMinidumpFileName);
 
+		// Note: std::wofstream won't throw an error on failure. The stream is merely marked as
+		// failed. We could check the state, and avoid the overhead of serializing the
+		// minidump if it failed, but it's hardly worth optimizing for an error case.
+		minidumpWos = GPOS_NEW(pmp) std::wofstream(szFileName);
+		minidumpOs = GPOS_NEW(pmp) COstreamBasic(minidumpWos.Pt());
+
+		mdmp.Init(minidumpOs.Pt());
```
Comment (oarap) [inline libgpopt/src/optimizer/COptimizer.cpp]: We can have this as constructor for CMiniDumperDXL.

File: libgpopt/include/gpopt/minidump/CSerializableQuery.h [hunk 2]
```diff
@@ -43,38 +43,23 @@ namespace gpopt
...
+			IMemoryPool *m_pmp;
 			
 			// private copy ctor
 			CSerializableQuery(const CSerializableQuery&);
 
 		public:
 
 			// ctor
-			CSerializableQuery(const CDXLNode *pdxlnQuery, const DrgPdxln *pdrgpdxlnQueryOutput, const DrgPdxln *pdrgpdxlnCTE);
+			CSerializableQuery(const CDXLNode *pdxlnQuery, const DrgPdxln *pdrgpdxlnQueryOutput, const DrgPdxln *pdrgpdxlnCTE, IMemoryPool *pmp);
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/minidump/CSerializableQuery.h]: In Orca, IMemoryPool is always the first variable.

File: libgpopt/include/gpopt/minidump/CSerializablePlan.h [hunk 3]
```diff
@@ -46,36 +46,23 @@ namespace gpopt
...
+			IMemoryPool *m_pmp;
+
 			// private copy ctor
 			CSerializablePlan(const CSerializablePlan&);
 
 		public:
 
 			// ctor
-			CSerializablePlan(const CDXLNode *pdxlnPlan, ULLONG ullPlanId, ULLONG ullPlanSpaceSize);
+			CSerializablePlan(const CDXLNode *pdxlnPlan, ULLONG ullPlanId, ULLONG ullPlanSpaceSize, IMemoryPool *pmp);
```
Comment (oarap) [inline libgpopt/include/gpopt/minidump/CSerializablePlan.h]: IMemoryPool is always the first variable

File: libgpopt/include/gpopt/minidump/CSerializableQuery.h [hunk 4]
```diff
@@ -43,38 +43,23 @@ namespace gpopt
 			// CTE DXL nodes
 			const DrgPdxln *m_pdrgpdxlnCTE;
 
-			CWStringDynamic *m_pstrQuery;
+			IMemoryPool *m_pmp;
```
Comment (oarap) [inline libgpopt/include/gpopt/minidump/CSerializableQuery.h]: this should go on top, first variable

File: libgpopt/include/gpopt/minidump/CSerializableQuery.h (hunk 2 above)
Comment (oarap) [inline libgpopt/include/gpopt/minidump/CSerializableQuery.h]: pmp, first parameter

File: server/src/unittest/gpopt/minidump/CMiniDumperDXLTest.cpp [hunk 5]
```diff
@@ -37,6 +37,8 @@
 #include "unittest/gpopt/translate/CTranslatorExprToDXLTest.h"
 #include "unittest/gpopt/CTestUtils.h"
 
+#include <fstream>
+
```
Comment (vraghavan78) [inline server/src/unittest/gpopt/minidump/CMiniDumperDXLTest.cpp]: can we avoid this?

File: libgpopt/src/minidump/CSerializableOptimizerConfig.cpp [hunk 6]
```diff
@@ -42,12 +42,13 @@ using namespace gpdxl;
...
 	(
-	const COptimizerConfig *poconf
+	const COptimizerConfig *poconf,
+	IMemoryPool *pmp
 	)
 	:
 	CSerializable(),
 	m_poconf(poconf),
-	m_pstrOptimizerConfig(NULL)
+	m_pmp(pmp)
```
Comment (oarap) [inline libgpopt/src/minidump/CSerializableOptimizerConfig.cpp]: goes top as the first param

File: libgpopt/src/minidump/CSerializableOptimizerConfig.cpp [hunk 7]
```diff
@@ -42,12 +42,13 @@ using namespace gpdxl;
 //---------------------------------------------------------------------------
 CSerializableOptimizerConfig::CSerializableOptimizerConfig
 	(
-	const COptimizerConfig *poconf
+	const COptimizerConfig *poconf,
+	IMemoryPool *pmp
```
Comment (oarap) [inline libgpopt/src/minidump/CSerializableOptimizerConfig.cpp]: make it first param

File: libgpopt/src/minidump/CSerializablePlan.cpp [hunk 8]
```diff
@@ -33,14 +33,16 @@ CSerializablePlan::CSerializablePlan
 	(
 	const CDXLNode *pdxlnPlan,
 	ULLONG ullPlanId,
-	ULLONG ullPlanSpaceSize
+	ULLONG ullPlanSpaceSize,
+	IMemoryPool *pmp
```
Comment (oarap) [inline libgpopt/src/minidump/CSerializablePlan.cpp]: make it first parameter

File: libgpopt/src/minidump/CSerializablePlan.cpp [hunk 9]
```diff
@@ -33,14 +33,16 @@ CSerializablePlan::CSerializablePlan
...
 	)
 	:
 	CSerializable(),
 	m_pdxlnPlan(pdxlnPlan),
 	m_pstrPlan(NULL),
 	m_ullPlanId(ullPlanId),
-	m_ullPlanSpaceSize(ullPlanSpaceSize)
+	m_ullPlanSpaceSize(ullPlanSpaceSize),
+	m_pmp(pmp)
```
Comment (oarap) [inline libgpopt/src/minidump/CSerializablePlan.cpp]: make it first param

File: libgpopt/src/minidump/CSerializableQuery.cpp [hunk 10]
```diff
@@ -33,14 +33,15 @@ CSerializableQuery::CSerializableQuery
 	(
 	const CDXLNode *pdxlnQuery, 
 	const DrgPdxln *pdrgpdxlnQueryOutput,
-	const DrgPdxln *pdrgpdxlnCTE
+	const DrgPdxln *pdrgpdxlnCTE,
+	IMemoryPool *pmp
```
Comment (oarap) [inline libgpopt/src/minidump/CSerializableQuery.cpp]: first param

File: libgpopt/src/minidump/CSerializableQuery.cpp [hunk 11]
```diff
@@ -33,14 +33,15 @@ CSerializableQuery::CSerializableQuery
...
 	)
 	:
 	CSerializable(),
 	m_pdxlnQuery(pdxlnQuery),
 	m_pdrgpdxlnQueryOutput(pdrgpdxlnQueryOutput),
 	m_pdrgpdxlnCTE(pdrgpdxlnCTE),
-	m_pstrQuery(NULL)
+	m_pmp(pmp)
```
Comment (oarap) [inline libgpopt/src/minidump/CSerializableQuery.cpp]: first param

File: libgpopt/src/optimizer/COptimizer.cpp [hunk 12]
```diff
@@ -199,28 +201,34 @@ COptimizer::PdxlnOptimize
 
 	BOOL fMinidump = GPOS_FTRACE(EopttraceMinidump);
 
+	// If minidump was requested, open the minidump file and initialize
+	// minidumper. (We create the minidumper object even if we're not
+	// dumping, but without the Init-call, it will stay inactive.)
 	CMiniDumperDXL mdmp(pmp);
+	CAutoP<std::wofstream> minidumpWos;
```
Comment (oarap) [inline libgpopt/src/optimizer/COptimizer.cpp]: not very Hungarian

File: libgpopt/src/optimizer/COptimizer.cpp [hunk 13]
```diff
@@ -199,28 +201,34 @@ COptimizer::PdxlnOptimize
...
+
+		mdmp.Init(minidumpOs.Pt());
+	}
 	CDXLNode *pdxlnPlan = NULL;
 	CErrorHandlerStandard errhdl;
 	GPOS_TRY_HDL(&errhdl)
 	{
 		CSerializableStackTrace serStack;
-		CSerializableOptimizerConfig serOptConfig(poconf);
+		CSerializableOptimizerConfig serOptConfig(poconf, pmp);
```
Comment (oarap) [inline libgpopt/src/optimizer/COptimizer.cpp]: pmp should go first

File: libgpopt/src/optimizer/COptimizer.cpp [hunk 14]
```diff
@@ -199,28 +201,34 @@ COptimizer::PdxlnOptimize
...
-		CSerializableQuery serQuery(pdxlnQuery, pdrgpdxlnQueryOutput, pdrgpdxlnCTE);
-		
-		if (fMinidump)
-		{
-			// pre-serialize query and optimizer configurations
-			serStack.AllocateBuffer(pmp);
-			serOptConfig.Serialize(pmp);
-			serQuery.Serialize(pmp);
-		}
+		CSerializableQuery serQuery(pdxlnQuery, pdrgpdxlnQueryOutput, pdrgpdxlnCTE, pmp);
```
Comment (oarap) [inline libgpopt/src/optimizer/COptimizer.cpp]: pmp should go first

File: libgpopt/src/optimizer/COptimizer.cpp [hunk 15]
```diff
@@ -268,9 +276,8 @@ COptimizer::PdxlnOptimize
 
 			if (fMinidump)
 			{
-				CSerializablePlan serPlan(pdxlnPlan, poconf->Pec()->UllPlanId(), poconf->Pec()->UllPlanSpaceSize());
-				serPlan.Serialize(pmp);
-				CMinidumperUtils::Finalize(&mdmp, true /* fSerializeErrCtxt*/, ulSessionId, ulCmdId, szMinidumpFileName);
+				CSerializablePlan serPlan(pdxlnPlan, poconf->Pec()->UllPlanId(), poconf->Pec()->UllPlanSpaceSize(), pmp);
```
Comment (oarap) [inline libgpopt/src/optimizer/COptimizer.cpp]: pmp should go first

File: libgpos/server/src/unittest/gpos/error/CMiniDumperTest.cpp [hunk 16]
```diff
@@ -61,7 +61,10 @@ CMiniDumperTest::EresUnittest_Basic()
 	IMemoryPool *pmp = amp.Pmp();
 
 	CMiniDumperStream mdrs(pmp);
-	mdrs.Init();
+
+	CWStringDynamic minidumpstr(pmp);
```
Comment (oarap) [inline libgpos/server/src/unittest/gpos/error/CMiniDumperTest.cpp]: not Hungarian

File: libgpos/server/src/unittest/gpos/error/CMiniDumperTest.cpp [hunk 17]
```diff
@@ -112,8 +100,10 @@ CMiniDumperTest::EresUnittest_Concurrency()
 
 	CWorkerPoolManager *pwpm = CWorkerPoolManager::Pwpm();
 
+	CWStringDynamic minidumpstr(pmp);
```
Comment (oarap) [inline libgpos/server/src/unittest/gpos/error/CMiniDumperTest.cpp]: not Hungarian

File: libnaucrates/src/xml/CXMLSerializer.cpp [hunk 18]
```diff
@@ -422,45 +422,41 @@ CXMLSerializer::WriteEscaped
...
 	
 	for (ULONG ulA = 0; ulA < ulLength; ulA++)
 	{
 		const WCHAR wc = wsz[ulA];
 		
-		const WCHAR *wszEscaped = NULL;
-		
-		for (ULONG ulB = 0; ulB < ulSpecialChars; ulB++)
+		switch (wc)
 		{
```
Comment (vraghavan78) [inline libnaucrates/src/xml/CXMLSerializer.cpp]: why can't we use clib::IWcsNCmp. I know it is equivalent but.

**Comment by vraghavan78** [comment]
This also needs GPDB changes correct.

**Comment by bhuvnesh2703** [comment]
@hlinnaka is there a gpdb pr corresponding to this? As we refer some of the modified functions in gpdb?

**Comment by hlinnaka** [comment]
On 04/10/2017 10:33 PM, Bhuvnesh wrote:
> @hlinnaka is there a gpdb pr corresponding to this? As we refer some
> of the modified functions in gpdb?

Yes, GPDB requires changes due to the changed interface functions. 
Opened PR for it now: https://github.com/greenplum-db/gpdb/pull/2197.

- Heikki

**Comment by vraghavan78** [comment]
Folk Orca version needs to change. Master is on 2.21

**Comment by oarap** [comment]
@hlinnaka I was going to merge this in, but there are a lot of failures in the debug test which are segfaults.

**Comment by hlinnaka** [comment]
On 04/21/2017 09:55 PM, Omer Arap wrote:
> @hlinnaka I was going to merge this in, but there are a lot of
> failures in the debug test which are segfaults.

Thanks! With unmodified master-branch, in debug mode, I'm seeing these 
failures:

97% tests passed, 3 tests failed out of 118

Total Test time (real) = 1034.06 sec

The following tests FAILED:
	 30 - gpos_test_CMemoryPoolBasicTest (OTHER_FAULT)
	 33 - gpos_custom_alloc_test_CMemoryPoolBasicTest (OTHER_FAULT)
	 42 - gpos_test_CSpinlockTest (OTHER_FAULT)
Errors while running CTest


I pushed a few commits to fix the extra failures caused by this PR, but 
those three unrelated ones remain. Please have a look.

- Heikki

**Comment by vraghavan78** [comment]
thanks @hlinnaka we will take a look

**Comment by oarap** [comment]
@hlinnaka This looks ok, I am not sure why debug is giving failures in your env. It is all passing in here. Merging this

**Pull request merged**